File: Backend/dno/src/main/java/com/catsnis/dno/service/ArchiveService.java
package com.catsnis.dno.service;

import com.catsnis.dno.common.utils.SecurityUtils;
import com.catsnis.dno.dto.ArchiveRequest;
import com.catsnis.dno.dto.ArchiveUpdateRequest;
import com.catsnis.dno.dto.ArchiveResponse;
import com.catsnis.dno.entity.Archive;
import com.catsnis.dno.entity.Archive.TypeArchive;
import com.catsnis.dno.entity.Archive.CategorieArchive;
import com.catsnis.dno.repository.ArchiveRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class ArchiveService {

    private final ArchiveRepository archiveRepository;
    private final SecurityUtils     securityUtils;   // ✅ injection pour archivedBy auto

    // ── Résoudre le nom complet de l'utilisateur courant ─────────────────────
    private String currentUserFullName() {
        return securityUtils.getCurrentUser()
                .map(p -> (p.getFirstName() + " " + p.getLastName()).trim())
                .orElse("Système");
    }

    // ── Upload fichier scanné ─────────────────────────────────────────────────
    // ✅ Stocké en BDD (BLOB) — le disque Render free tier est éphémère.
    @Transactional
    public ArchiveResponse uploadScanne(MultipartFile file, ArchiveRequest dto) throws IOException {
        byte[] bytes = file.getBytes();

        Archive archive = Archive.builder()
                .titre(dto.getTitre())
                .type(TypeArchive.SCANNE)
                .categorie(dto.getCategorie())
                .fileName(file.getOriginalFilename())
                .data(bytes)
                .fileSize((long) bytes.length)
                .mimeType(file.getContentType())
                .description(dto.getDescription())
                .archivedBy(currentUserFullName())   // ✅ auto depuis le token JWT
                .relatedId(dto.getRelatedId())
                .relatedCode(dto.getRelatedCode())
                .build();

        return toResponse(archiveRepository.save(archive));
    }

    // ── Archiver automatiquement un document imprimé (sans fichier) ──────────
    @Transactional
    public ArchiveResponse archiverImprime(ArchiveRequest dto) {
        Archive archive = Archive.builder()
                .titre(dto.getTitre())
                .type(TypeArchive.IMPRIME)
                .categorie(dto.getCategorie() != null ? dto.getCategorie() : CategorieArchive.AUTRE)
                .description(dto.getDescription())
                .archivedBy(currentUserFullName())   // ✅ auto depuis le token JWT
                .relatedId(dto.getRelatedId())
                .relatedCode(dto.getRelatedCode())
                .build();

        return toResponse(archiveRepository.save(archive));
    }

    // ── Archiver (ou remplacer) le PDF généré automatiquement ─────────────────
    //    pour une intervention ou un déploiement. Stocké en BDD (BLOB) —
    //    survit aux redéploiements, contrairement à un fichier sur disque.
    //
    //    Dédoublonnage : si une archive existe déjà pour ce relatedId +
    //    categorie + type=IMPRIME (le PDF a déjà été généré une première fois),
    //    on remplace le contenu existant au lieu d'en créer un nouveau —
    //    important car generateInterventionPdf/generateDeploymentPdf peuvent
    //    être appelées à chaque téléchargement, pas uniquement à la création.
    @Transactional
    public ArchiveResponse archiverPdfGenere(byte[] pdfBytes, String titre,
                                             CategorieArchive categorie,
                                             Long relatedId, String relatedCode) {
        String fileName = titre.replaceAll("[^a-zA-Z0-9-_]", "_") + ".pdf";

        Archive archive = archiveRepository
                .findFirstByRelatedIdAndCategorieAndType(relatedId, categorie, TypeArchive.IMPRIME)
                .orElse(null);

        if (archive != null) {
            archive.setTitre(titre);
            archive.setFileName(fileName);
            archive.setData(pdfBytes);
            archive.setFileSize((long) pdfBytes.length);
            archive.setMimeType("application/pdf");
            archive.setArchivedBy(currentUserFullName());
            archive.setRelatedCode(relatedCode);
        } else {
            archive = Archive.builder()
                    .titre(titre)
                    .type(TypeArchive.IMPRIME)
                    .categorie(categorie)
                    .fileName(fileName)
                    .data(pdfBytes)
                    .fileSize((long) pdfBytes.length)
                    .mimeType("application/pdf")
                    .archivedBy(currentUserFullName())
                    .relatedId(relatedId)
                    .relatedCode(relatedCode)
                    .build();
        }

        return toResponse(archiveRepository.save(archive));
    }

    // ── Mettre à jour les métadonnées uniquement ──────────────────────────────
    @Transactional
    public ArchiveResponse update(Long id, ArchiveUpdateRequest dto) {
        Archive archive = archiveRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Archive introuvable : " + id));

        if (dto.getTitre() != null && !dto.getTitre().isBlank())
            archive.setTitre(dto.getTitre().trim());
        if (dto.getDescription() != null)
            archive.setDescription(dto.getDescription().trim().isEmpty()
                    ? null : dto.getDescription().trim());
        if (dto.getCategorie() != null)
            archive.setCategorie(dto.getCategorie());
        if (dto.getRelatedCode() != null)
            archive.setRelatedCode(dto.getRelatedCode().trim().isEmpty()
                    ? null : dto.getRelatedCode().trim());
        if (dto.getRelatedId() != null)
            archive.setRelatedId(dto.getRelatedId());

        // ✅ Mettre à jour archivedBy avec l'utilisateur qui fait la modification
        archive.setArchivedBy(currentUserFullName());

        return toResponse(archiveRepository.save(archive));
    }

    // ── Mettre à jour avec remplacement du fichier (SCANNE uniquement) ────────
    @Transactional
    public ArchiveResponse updateWithFile(Long id, MultipartFile file, ArchiveUpdateRequest dto)
            throws IOException {
        Archive archive = archiveRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Archive introuvable : " + id));

        if (archive.getType() != TypeArchive.SCANNE)
            throw new RuntimeException("Seuls les documents scannés peuvent avoir un fichier.");

        byte[] bytes = file.getBytes();
        archive.setFileName(file.getOriginalFilename());
        archive.setData(bytes);
        archive.setFileSize((long) bytes.length);
        archive.setMimeType(file.getContentType());

        if (dto.getTitre() != null && !dto.getTitre().isBlank())
            archive.setTitre(dto.getTitre().trim());
        if (dto.getDescription() != null)
            archive.setDescription(dto.getDescription().trim().isEmpty()
                    ? null : dto.getDescription().trim());
        if (dto.getCategorie() != null)
            archive.setCategorie(dto.getCategorie());
        if (dto.getRelatedCode() != null)
            archive.setRelatedCode(dto.getRelatedCode().trim().isEmpty()
                    ? null : dto.getRelatedCode().trim());
        if (dto.getRelatedId() != null)
            archive.setRelatedId(dto.getRelatedId());

        // ✅ Mettre à jour archivedBy avec l'utilisateur qui fait la modification
        archive.setArchivedBy(currentUserFullName());

        return toResponse(archiveRepository.save(archive));
    }

    // ── Liste paginée avec filtres ────────────────────────────────────────────
    public Page<ArchiveResponse> list(TypeArchive type, CategorieArchive categorie,
                                      String keyword, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        return archiveRepository.search(
                type,
                categorie,
                (keyword != null && !keyword.isBlank()) ? keyword : null,
                pageable
        ).map(this::toResponse);
    }

    // ── Télécharger un fichier ────────────────────────────────────────────────
    // ✅ Servi directement depuis la BDD (BLOB), plus depuis le disque.
    public Resource download(Long id) {
        Archive archive = archiveRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Archive introuvable : " + id));
        if (archive.getData() == null)
            throw new RuntimeException("Aucun fichier associé à cette archive");
        return new ByteArrayResource(archive.getData());
    }

    // ── Nom fichier pour l'entête Content-Disposition ─────────────────────────
    public String getFileName(Long id) {
        return archiveRepository.findById(id)
                .map(a -> a.getFileName() != null ? a.getFileName() : "archive_" + id + ".bin")
                .orElse("archive.bin");
    }

    // ── Supprimer ─────────────────────────────────────────────────────────────
    @Transactional
    public void delete(Long id) {
        archiveRepository.deleteById(id);
    }

    // ── Stats rapides ─────────────────────────────────────────────────────────
    public Map<String, Long> stats() {
        return Map.of(
                "total",         archiveRepository.count(),
                "imprimes",      archiveRepository.countByType(TypeArchive.IMPRIME),
                "scannes",       archiveRepository.countByType(TypeArchive.SCANNE),
                "interventions", archiveRepository.countByCategorie(CategorieArchive.INTERVENTION),
                "deploiements",  archiveRepository.countByCategorie(CategorieArchive.DEPLOIEMENT)
        );
    }

    // ── Mapper ────────────────────────────────────────────────────────────────
    private ArchiveResponse toResponse(Archive a) {
        ArchiveResponse r = new ArchiveResponse();
        r.setId(a.getId());
        r.setTitre(a.getTitre());
        r.setType(a.getType());
        r.setCategorie(a.getCategorie());
        r.setFileName(a.getFileName());
        r.setFileSize(a.getFileSize());
        r.setMimeType(a.getMimeType());
        r.setDescription(a.getDescription());
        r.setArchivedBy(a.getArchivedBy());
        r.setArchivedAt(a.getArchivedAt());
        r.setRelatedId(a.getRelatedId());
        r.setRelatedCode(a.getRelatedCode());
        if (a.getFileName() != null)
            r.setDownloadUrl("/api/archives/download/" + a.getId());
        return r;
    }
}

File: Backend/catusnis-front/src/services/archiveService.ts
import api from './api';

// ── Types alignés sur le backend ─────────────────────────────────────────────
export type TypeArchive      = 'IMPRIME' | 'SCANNE';
export type CategorieArchive = 'INTERVENTION' | 'DEPLOIEMENT' | 'ACQUISITION' | 'BOOKLET' | 'AUTRE';

export interface ArchiveResponse {
  id:          number;
  titre:       string;
  type:        TypeArchive;
  categorie:   CategorieArchive;
  fileName:    string | null;
  fileSize:    number | null;
  mimeType:    string | null;
  description: string | null;
  archivedBy:  string | null;
  archivedAt:  string;
  relatedId:   number | null;
  relatedCode: string | null;
  downloadUrl: string | null;
}

export interface ArchiveRequest {
  titre:        string;
  type:         TypeArchive;
  categorie:    CategorieArchive;
  description?: string;
  archivedBy?:  string;
  relatedId?:   number;
  relatedCode?: string;
}

export interface ArchiveStats {
  total:         number;
  imprimes:      number;
  scannes:       number;
  interventions: number;
  deploiements:  number;
}

export interface SpringPage<T> {
  content:       T[];
  totalElements: number;
  totalPages:    number;
  number:        number;
  size:          number;
}

// ── Constantes formats acceptés (centralisées ici pour cohérence) ─────────────
export const ACCEPTED_EXTENSIONS = ['pdf', 'doc', 'docx', 'zip', 'rar', '7z'] as const;
export type  AcceptedExtension   = typeof ACCEPTED_EXTENSIONS[number];

export const ACCEPT_MIME =
  '.pdf,.doc,.docx,.zip,.rar,.7z,' +
  'application/pdf,' +
  'application/msword,' +
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document,' +
  'application/zip,' +
  'application/x-zip-compressed,' +
  'application/x-rar-compressed,' +
  'application/vnd.rar,' +
  'application/x-7z-compressed';

export const MAX_SIZE_MB = 30;
export const MAX_SIZE_B  = MAX_SIZE_MB * 1024 * 1024;

/** Valide l'extension et la taille. Retourne un message d'erreur ou null. */
export const validateFile = (file: File): string | null => {
  if (file.size > MAX_SIZE_B)
    return `Fichier trop volumineux. Taille max : ${MAX_SIZE_MB} Mo.`;
  const ext = file.name.split('.').pop()?.toLowerCase() as AcceptedExtension | undefined;
  if (!ext || !ACCEPTED_EXTENSIONS.includes(ext))
    return `Format non supporté. Formats acceptés : ${ACCEPTED_EXTENSIONS.join(', ').toUpperCase()}.`;
  return null;
};

/** Retourne l'icône Bootstrap et la couleur selon le type MIME ou l'extension. */
export const getFileIcon = (
  mimeType?: string | null,
  fileName?: string | null,
): { icon: string; color: string } => {
  const ext = fileName?.split('.').pop()?.toLowerCase();
  if (mimeType === 'application/pdf'  || ext === 'pdf')  return { icon: 'bi-file-earmark-pdf-fill',  color: 'danger'   };
  if (mimeType?.includes('word')      || ext === 'doc' || ext === 'docx')
                                                          return { icon: 'bi-file-earmark-word-fill', color: 'primary'  };
  if (ext === 'zip' || mimeType?.includes('zip'))         return { icon: 'bi-file-zip-fill',          color: 'warning'  };
  if (ext === 'rar' || mimeType?.includes('rar'))         return { icon: 'bi-file-zip-fill',          color: 'orange'   };
  if (ext === '7z'  || mimeType?.includes('7z'))          return { icon: 'bi-file-zip-fill',          color: 'secondary'};
  return { icon: 'bi-file-earmark-fill', color: 'secondary' };
};

const ArchiveService = {

  // ── Liste paginée avec filtres ──────────────────────────────────────────────
  list: async (
    page      = 0,
    size      = 12,
    type?:      TypeArchive,
    categorie?: CategorieArchive,
    keyword?:   string,
  ): Promise<SpringPage<ArchiveResponse>> => {
    const params: Record<string, any> = { page, size };
    if (type)      params.type      = type;
    if (categorie) params.categorie = categorie;
    if (keyword)   params.keyword   = keyword;
    const res = await api.get<SpringPage<ArchiveResponse>>('/api/archives', { params });
    return res.data;
  },

  // ── Upload document scanné ──────────────────────────────────────────────────
  uploadScanne: async (file: File, dto: ArchiveRequest): Promise<ArchiveResponse> => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('data', new Blob([JSON.stringify(dto)], { type: 'application/json' }));
    const res = await api.post<ArchiveResponse>('/api/archives/upload', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return res.data;
  },

  // ── Archiver un document imprimé (sans fichier) ─────────────────────────────
  archiverImprime: async (dto: ArchiveRequest): Promise<ArchiveResponse> => {
    const res = await api.post<ArchiveResponse>('/api/archives/imprime', dto);
    return res.data;
  },

  // ── Mettre à jour les métadonnées uniquement ────────────────────────────────
  update: async (id: number, dto: Partial<ArchiveRequest>): Promise<ArchiveResponse> => {
    const res = await api.put<ArchiveResponse>(`/api/archives/${id}`, dto);
    return res.data;
  },

  // ── Mettre à jour avec remplacement du fichier ──────────────────────────────
  updateWithFile: async (id: number, file: File, dto: Partial<ArchiveRequest>): Promise<ArchiveResponse> => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('data', new Blob([JSON.stringify(dto)], { type: 'application/json' }));
    const res = await api.put<ArchiveResponse>(`/api/archives/${id}/upload`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return res.data;
  },

  // ── Télécharger un fichier ──────────────────────────────────────────────────
  download: async (id: number, fileName: string): Promise<void> => {
    const res = await api.get(`/api/archives/download/${id}`, { responseType: 'blob' });
    const url  = URL.createObjectURL(new Blob([res.data]));
    const link = document.createElement('a');
    link.href     = url;
    link.download = fileName || `archive_${id}`;
    link.click();
    URL.revokeObjectURL(url);
  },

  // ── Supprimer ───────────────────────────────────────────────────────────────
  delete: async (id: number): Promise<void> => {
    await api.delete(`/api/archives/${id}`);
  },

  // ── Stats ───────────────────────────────────────────────────────────────────
  stats: async (): Promise<ArchiveStats> => {
    const res = await api.get<ArchiveStats>('/api/archives/stats');
    return res.data;
  },
};

export default ArchiveService;